### PR TITLE
DRAFT: Prad multiple field grids

### PR DESCRIPTION
**Pull request by pheuer**
My attempt to address #1725 

Wait to merge until #1274 is done, since that will refactor the structure of the charged particle radiography code. 

TODO list

- [x] Allow a list of 'grid' objects to be passed into the 'grid' argument. If only one grid is passed, wrap it in a list so that there is always a list. 

- [x] Modify  max theta hit grid for each grid in the list, then take the maximum

- [x] Similarly, for the rest of the `__init__ method`, modify so that all the validation etc. for the grids is being run on each grid in the list. 

- [x] Modify `_coast_to_grid` to calculate `dist` to each grid and then coast to the minimum distance (e.g. the edge of the largest grid). 

- [x] Within `_push`, run `on_grid` on each of the grids. `self.on_grid` will need to be a list of lists now, recording which particles are on each grid. NOTE: -> now on_grid has shape [nparticles, ngrids]

- [x] (The big one): the interpolation section of  `_push` needs to be modified so that you interpolate the correct fields for the correct particles and then sum those fields as I proposed above.  NOTE -> We interpolate every particle on every grid, and just add it all together for each particles. So grids are additive.

- [x] Modify `_adaptive_dt` to calculate the grid resolution of each grid _that currently has particles on it_  based on the ongrid method. We'll need a different dt for each grid. The timestep for each particle needs to be governed by the highest resolution grid it is on. E.g. if it is on both, the smaller dt wins. To avoid unnecessary runtime, we should probably calculate dt for each particle separately (so that those on larger grids can move faster than those on smaller grids). NOTE -> Now calculating a separate dt for every particle, based on what grids it is on.  When a particle is on no grid, give it the largest gridstep as dt.

- [x] Modify `_push` to push the particles in batches based on which timestep they have. If we push everything based on the smaller timestep, the particles will take a very long time to move through the large length scale low res grid. NOTE -> We now handle this by having every particle have its own dt.

- [x] Modify `_stop_condition` as necessary to deal with changes to `entered_grid`. We want to know what fraction of the particle have entered A grid and are now on NO grid. 

- [x] Implement at least one test with two grids of different scales and an expected analytical outcome and demonstrate that the new functionality works. That will go in this file: https://github.com/PlasmaPy/PlasmaPy/blob/main/plasmapy/diagnostics/tests/test_charged_particle_radiography.py

### CODE REVIEW

**Comment by pheuer** [comment]
@weipengyao I took a stab at this and it seems to be working - give it a try (using this branch) and let me know what you think!

**Comment by weipengyao** [comment]
> @weipengyao I took a stab at this and it seems to be working - give it a try (using this branch) and let me know what you think!

@pheuer Hi Peter, thanks again for the updates! Yes, it works! Now I am playing with the size, resolution, and the profile of the additional grid. But the function itself is good, and I do see a difference in the final results with and w/o the 2nd grid.

File: plasmapy/diagnostics/charged_particle_radiography.py [hunk 1]
```diff
@@ -733,56 +755,78 @@ def load_particles(
...
-        dist = np.min(np.linalg.norm(self.grid_arr - self.source, axis=3))
+        # Distance from the source to the nearest point on any grid
+        dist = np.min(
+            np.array(
+                [
+                    np.min(np.linalg.norm(arr - self.source, axis=3))
+                    for arr in self.grids_arr
+                ]
+            )
+        )
```
Comment (StanczakDominik) [inline plasmapy/diagnostics/charged_particle_radiography.py:820]: Unless you have a WHOLE LOT of grids, there's not going to be much of a gain in having the list of grids be an array as well, so we can simplify here:
```suggestion
        dist = min(
                    np.min(np.linalg.norm(arr - self.source, axis=3))
                    for arr in self.grids_arr
        )
```

**Review by StanczakDominik — COMMENTED** [review]
Looks good! Pointed out a few possible rewrites, and the test automation issue seems challenging, but... yeah, looks good!

File: plasmapy/diagnostics/charged_particle_radiography.py (hunk 1 above)
Comment (StanczakDominik) [reply plasmapy/diagnostics/charged_particle_radiography.py:820]: If we were reusing the array of minima along axis=3 sure, but these are passed straight to min. So you're just adding an iteration over all the elements to change their container before iterating over them again to find the scalar minimum, unless I'm misunderstanding this? Try it in ipython with a %timeit!

Comment (pheuer) [reply plasmapy/diagnostics/charged_particle_radiography.py:820]: Got it - thanks!

**Comment by pheuer** [comment]
@StanczakDominik I've updated from main and merged with the changes from #1274 , and made the changes you suggested above. Writing an automated test to ensure this works would be heavy (require a high resolution Tracker run) and I don't really think it's worth it: if it runs to completion without error it should be working correctly.

**Comment by StanczakDominik** [comment]
I'm happy to skip it, then!

I'll just figure out the doc errors and we can probably merge it. I'll take it from here :)

**Comment by pheuer** [comment]
> I'll just figure out the doc errors and we can probably merge it. I'll take it from here :)

Doc issue resolved! It was easy, just a stupid typo

**Review by StanczakDominik — APPROVED** [review]

**Comment by pheuer** [comment]
Awesome, thanks @StanczakDominik

**Comment by StanczakDominik** [comment]
Nah, thank you for doing this and for finding the doc bug ;)

**Pull request merged**